pode/utils.py

```diff
@@ -1,100 +1,233 @@
-    """
-    'If and only if' - biconditional logical connective.
-    Returns `True` if both statements are `True` or `False`.
-    """
-    return not (statement ^ other_statement)
-
-
-@lru_cache(maxsize=None)
-def cached_sum(*values: Tuple[float]) -> float:
-    """
-    Calculates sum of input arguments
-    and remembers results for the same input
-    """
-    return sum(values)
-
-
-def starfilter(predicate: Predicate,
-               iterable: Iterable[T]) -> Iterator[T]:
-    """
-    Can be used when elements of an iterable are packed in tuples
-    but the predicate function expects separate arguments.
-    Similar to `itertools.starmap`.
-    """
-    return filter(pack(predicate), iterable)
+        segment_to_start = Segment(vertices[start_index], start)
+        segment_to_end = Segment(vertices[start_index], end)
+        if segment_to_start.length < segment_to_end.length:
+            return [start, end]
+        else:
+            result = vertices[start_index + 1:-1] + vertices[:end_index + 1]
+    head = [vertices[start_index]] if start in vertices else [start]
+    tail = [] if end in vertices else [end]
+    return head + result + tail
+
+
+def edges(contour: Contour) -> Iterator[Segment]:
+    vertices = [*contour.vertices, contour.vertices[0]]
+    yield from (Segment(vertices[index], vertices[index + 1])
+                for index in range(len(vertices) - 1))
+
+
+def shrink_collinear_vertices(contour: Contour) -> List[Point]:
+    vertices = list(contour.vertices)
+    index = -len(vertices) + 1
+    while index < 0:
+        while (max(2, -index) < len(vertices)
+               and (Contour([vertices[index + 2],
+                             vertices[index + 1],
+                             vertices[index]]).orientation
+                    is Orientation.COLLINEAR)):
+            del vertices[index + 1]
+        index += 1
+    while index < len(vertices):
+        while (max(2, index) < len(vertices)
+               and (Contour([vertices[index - 2],
+                             vertices[index - 1],
+                             vertices[index]]).orientation
+                    is Orientation.COLLINEAR)):
+            del vertices[index - 1]
+        index += 1
+    return vertices
+
+
+@singledispatch
+def to_fractions(geometry: GeometryType) -> GeometryType:
+    raise TypeError(f"Unsupported type: {type(geometry)}")
+
+
+@to_fractions.register
+def _(geometry: Point) -> Point:
+    return Point(Fraction(geometry.x), Fraction(geometry.y))
+
+
+@to_fractions.register
+def _(geometry: Contour) -> Contour:
+    return Contour(list(map(to_fractions, geometry.vertices)))
+
+
+@to_fractions.register
+def _(geometry: Polygon) -> Polygon:
+    return Polygon(to_fractions(geometry.border),
+                   list(map(to_fractions, geometry.holes)))
+
+
+def union(*geometries: Shaped) -> Union[Shaped, Empty]:
+    if not geometries:
+        return EMPTY
+    return reduce(operator.or_, geometries)

```

Comment:
this can be one-liner with useful [`initializer` parameter](https://docs.python.org/3/library/functools.html#functools.reduce)
```python
def union(*geometries: Shaped) -> Union[Shaped, Empty]:
    return reduce(operator.or_, geometries, EMPTY)
```

Reply:
That's cool! I completely forgot about it!